server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java

```diff
@@ -343,4 +357,82 @@ public void collect(GarbageCollectionEnvironment gce)
       deleteConfirmed(gce, candidateMap);
     }
   }
+
+  /**
+   * Track metadata rows read to help validate that gc scan has complete information to make a
+   * decision on deleting files
+   */
+  private static class MetadataReadTracker implements AutoCloseable {
```

Comment:
This tracker class is now a substantial portion of the algorithm class file. I would consider moving this class out to its own separate file in the same package. Since it's a static class and shouldn't be tightly coupled to any of the algorithm class' state, this should be easy to do.

Reply:
The tracker class is not expected to be used outside of the GarbageCollectionAlgorithm - it is not a general class, more of an implementation detail.  Moving it outside of the GCA seems to imply that other classes might want to use it.  Currently, it was not designed for anything else other than tracking the metadata seen in the context of the GCA metadata scan.  If other classes need similar functionality, then this could be refactored to a utility somewhere?  But until that need is identified, keeping it internal to GCA seems proper.

Reply:
In 3bba72af51 the class was renamed MetadataRowReadTracker.  Looking at porting the check to 2.1, Ample changes / simplifies the check and this class is not needed to track reading a row-at-at-time.  Because this is a one-off and will not be carried forward, I favor keeping this enclosed with the GCA for now.